### PR TITLE
Add $track/$undo helper

### PR DESCRIPTION
**Pull request by KevinBatdorf**
It's working but I want to do some more testing and ~~write some feature tests.~~

~~Also, I refactored quite a bit out of `$component` so I'll probably move that to another PR.~~ (#24)

There's no `$redo` either which might be nice. Will look into what's involved for that as well. 

Demo is at: http://localhost:3000/examples/undo.html

### CODE REVIEW

File: src/undo.js
Comment (KevinBatdorf) [reply src/undo.js:33]: Did you try it? I couldn't get the UI to update without it. I tried to comment out each updateElements call inidividually but then it would break. Was thinking maybe a race condition with how the deep-diff applies the patch. Could possibly instead proxy `$el.previousComponentState` or something, then updateElements in the set trap. Will mess around more after writing some tests.

Comment (KevinBatdorf) [reply src/undo.js:33]: Welp, I removed everything now and it works. May have been needed in an early iteration or something. Seems ok without it though.

Comment (KevinBatdorf) [reply src/undo.js:33]: Testing some more now, with them removed the component isn't updated when `$el.changes` is empty, so the undo button will still display. Will look into  another approach though.

**Comment by KevinBatdorf** [comment]
@SimoTod Was able to clean it up greatly with your feedback and seems to work now just the same. Any thought about additional tests to add?

File: src/undo.js
Comment (SimoTod) [reply src/undo.js]: I see, do we wan't to make $el.changes reactive though? 
It's not part of x-data so, technically speaking, it should neither be reactive nor used on the UI.

Comment (KevinBatdorf) [reply src/undo.js]: I think there should be a way for users to know whether an "undo" operation is possible. For example, you might disable or hide an undo button when there are no available changes. Same goes for redo (if that gets implemented).

The current version in this PR works though with that undo functionality. I just had to restructure a few things.

**Comment by SimoTod** [comment]
Not necessary for this PR but it seems we are setting a lot of custom properties on DOM elements in our helpers (not just this one), I just wanted to point out that, especially if we use simple names as `changes` any other libraries can use them and generate conflicts/bugs.
Maybe we should think about using a namespace such as `$el.__alpine_collective.*`

Also, does it work okay with Livewire? I think Livewire overrides the DOM element using clone and it does not call x-init so, after a Livewire roundtrip, stuff like $el.changes will be lost (it's not a problem for parent since it's just a performance improvement there and we recalculate it if needed).

I don't think we need more tests at this stage.

**Comment by KevinBatdorf** [comment]
Namespacing to `$el.__alpine_collective` is a good idea, or maybe something shorter like `$el.__xc.`?

> Also, does it work okay with Livewire? 

Haven't tested it but can when I get some time. Any suggestions if not?

> I think Livewire overrides the DOM element using clone and it does not call x-init

That would seem like a bug right?

**Comment by SimoTod** [comment]
> Namespacing to `$el.__alpine_collective` is a good idea, or maybe something shorter like `$el.__xc.`?

yeah, as long as it something weird that it's unlikely to be used by other libraries, I'm done with that.

The livewire thing definitely needs testing, I might be mistaken. Not sure what we should do, it it does not work though.

**Comment by KevinBatdorf** [comment]
Maybe just store the changes on `Window`? Or maybe detect Livewire first then do something extra

File: src/undo.js
Comment (SimoTod) [reply src/undo.js]: We could have $hasChanges which compares the current $el.__x.unosservedData with safeEval($el.getAttribute('x-data')), but maybe it's an overComplication.

It's just that, if you have a computed property counting the number of refreshes, I suspect it will be increased by 2 every time

```javascript
function data() {
    let counter = 0
    return {
       foo: 'bar',
       init() {this.$track()}
       getRefreshes() { return ++counter()}
    }
}
```

**Comment by KevinBatdorf** [comment]
Update now to require the user add a `__xhistory` property:

```html
<div x-data="{ number: 0, __xhistory: [] }"></div>
```
This works a bit better than before too as it will scope properties you pass into `$track()` (still wip)

@SimoTod I haven't tested it with LiveWire yet, but with this approach it should work since the property is on the component directly, right?

**Comment by SimoTod** [comment]
@KevinBatdorf Yeah, it should work since it will get cloned by livewire.

Do we need to ask users to create that property though? I think we can check if the property exists and if, not, instantiate it to an empty array when we use $track. To avoid accidental collisions, I would call it `$history` and document it as a feature so everyone knows we are using it. For the final user, there's no difference between a magic helper and a real data field since they'll get resolved In the same way by Alpine, and people kind of know that they shouldn't define properties starting with $ because they could conflict with Alpine internals. What do you think?

**Comment by KevinBatdorf** [comment]
@SimoTod Would it be too late if we set it for them? For example, if they put it in `x-show`

```html
x-show.transition.1000="$history.length"
```

**Comment by SimoTod** [comment]
Mmmh, maybe. I assume we don't know if they use $track in x-init or in an x-init callback.
But... We can add an $history helper which returns the $history property (creating it if it does not exists).
The only problem (?) is that people can alter the history

**Comment by KevinBatdorf** [comment]
I'll try them out. I agree it's always better for the user to do as little setup as possible. The $history helper might work.

> The only problem (?) is that people can alter the history.

I don't think this is an issue though since it would have to be intentional. And maybe there's even a use case to inject some fake history or something.

**Comment by KevinBatdorf** [comment]
Updated so it will all show under `$data` but at least namespaced, `$el.__x.$data.__xc.propertiesBeingWatched`, for example.

Tomorrow I'll see if I can write some `clone()` tests

**Comment by SimoTod** [comment]
Do you mean when you clone? That's fine, you will have defined your helper on the first iteration, you don't want to override them. I'll test it soon

**Comment by KevinBatdorf** [comment]
Yes, when you clone. It's the data that get's update on each mutation that won't carry over I'm assuming. Heading offline now though. Will check in tomorrow.

(had some issues merging too)

To compare without the refactor:
https://github.com/alpine-collective/alpine-magic-helpers/compare/chore/refactor-component-utils...feature/add-track-undo-helper

**Comment by SimoTod** [comment]
I couldn't follow really well the logic but it seems that there are too many collisions/hack we need to implements.
I've got a new possible strategy (simplified code):

```javascript
import {
    checkForAlpine,
    componentData,
    updateOnMutation,
    saferEval,
} from './utils'
import { DeepDiff } from 'deep-diff'

const history = new WeakMap()

const AlpineUndoMagicMethod = {
    start() {
        checkForAlpine()

        Alpine.addMagicProperty('history', function ($el) {
            return history.has($el.__x) ? history.get($el.__x) : []
        })

        Alpine.addMagicProperty('track', function ($el) {
            return function () {
                updateOnMutation($el, () => {
                    if (!history.has($el.__x)) {
                        history.set(
                            $el.__x,
                            {
                                previous: saferEval($el.getAttribute('x-data'), {}),
                                changes: [],
                                get length() {
                                    return this.changes.length
                                }
                            }
                        )
                    }
                    const fresh = componentData($el)
                    let changes = DeepDiff.diff(history.get($el.__x).previous, fresh, true)

                    if (changes) {
                        history.get($el.__x).previous = fresh
                        history.get($el.__x).changes.push(changes)
                        $el.__x.updateElements($el)
                    }
                })
            }
        })

        Alpine.addMagicProperty('undo', function ($el) {
            return function () {
                if (history.has($el.__x) && history.get($el.__x).changes.length > 0) {
                    const changes = history.get($el.__x).changes.pop()
                    console.log('Revert change with DeepDiff. and set history.get($el.__x).previous')
                }
            }
        })
    },
}

const alpine = window.deferLoadingAlpine || ((alpine) => alpine())

window.deferLoadingAlpine = function (callback) {
    alpine(callback)

    AlpineUndoMagicMethod.start()
}

export default AlpineUndoMagicMethod

```

**Comment by KevinBatdorf** [comment]
I thought about doing something like that but wouldn't it be sharing state with all components then? I'll give it a try.

**Comment by SimoTod** [comment]
It may be that we need to use `$el.__x.$data`, i can't remember what clone uses but it should work. being a weakmap it will also bin the entry when the component is removed (for stuff like turbolinks).

**Comment by SimoTod** [comment]
`$el.__x.unobservedData` actually

**Comment by KevinBatdorf** [comment]
Ok I'll give that a try. i'll try to write a test for it too. Otherwise I think it's looking pretty good

**Comment by KevinBatdorf** [comment]
Added a test which is failing. Tried using `history.set($el.__x` and `history.set($el.__x.unobservedData` but neither worked. I think because the latter would be the current component state rather than the initial state. I'll see if there's something else I can use.

**Comment by KevinBatdorf** [comment]
One initial idea would be to update Alpine where magic helpers are passed in:

https://github.com/alpinejs/alpine/blob/master/src/component.js#L25

If feels like a bug that the helper would only have access to the previous component and not the new clone. It could be passed in as a second parameter:

```js
Object.entries(Alpine.magicProperties).forEach(([name, callback]) => {
  Object.defineProperty(dataExtras, `$${name}`, {
    get: function () {
      // Added a second param here
      return callback(canonicalComponentElementReference, this.$el)
    }
  });
})
```
Then inside helpers we could just compare the two objects and if they don't match then we're dealing with a clone.

Just a thought. I'll keep trying other ideas.

**Comment by KevinBatdorf** [comment]
Updated my last comment a bit. I don't think will work unless we can get access to the newly cloned component somehow. 

When is `clone()` most commonly used? Could we just not support it (even if just for now)?

**Comment by KevinBatdorf** [comment]
I have this working now but only if I can get that update to Alpine core merged in. I'll send a PR probably tomorrow

ie. Updating to include the second param:
```js
return callback(canonicalComponentElementReference, this.$el)
```

It won't work on components that are cloned yet still exist (so two copies on a page) though as the events are still registered, thus `@click="$undo()"` would fire on both. I don't see a good way around it but I don't think that's a common use case anyway (and perhaps something to be improved on in Alpine itself if needed).

**Comment by SimoTod** [comment]
> It won't work on components that are cloned yet still exist (so two copies on a page)

That use case is not supported by alpine so I don't think it's a problem. It would mess up with the existing reactivity framework and introduce other bugs as well. 

Currently, clone is only an internal API to support livewire. When something changes in livewire, the backend rerenders part of the HTML. Without clone, Alpine components in that snippet would get reset every time.
Clone is just initializing those "new" component (since they replace the old elements) with the old data context so nothing gets lost.

Not sure why Alpine does `let canonicalComponentElementReference = componentForClone ? componentForClone.$el : this.$el`, it doesn't make much sense to me because I would expect it to always receive the current element.
We probably need to ask Caleb.

**Comment by SimoTod** [comment]
I tried `return callback(this.$el)` and `$wire` works correctly. 🤷‍♂️  I don't think we need 2 parameters, we could just ask Caleb if there is a reason why we pass the old HtmlElement.

**Comment by KevinBatdorf** [comment]
Yeah I'd be curious to know if there's a use for having the old element there. I could see it being useful if you could access properties on it before it's removed. It might be good to have it passed in, but maybe the stale component could be the second parameter instead. 

```js
return callback(this.$el, canonicalComponentElementReference)
```

**Comment by SimoTod** [comment]
not sure, I would wait for a real use case before overcomplicating it. The callback will run after the element is removed. 
At that point, it's either going to be undefined or, more likely, to create a memory leak since the garbage collector won't free the memory.

**Comment by KevinBatdorf** [comment]
Ah ok. Not sure how it's being used. In the [test for clone on Alpine](https://github.com/alpinejs/alpine/blob/c430b5a6856859b8fa2d786bdd99738822dfebe2/test/constructor.spec.js#L336) it actually creates a clone on a second object (so both exist at the same time), but I'm guessing in Livewire it's being used differently.

Do you want to open the PR about it to get Caleb's reply on it?

**Comment by SimoTod** [comment]
yeah, the test is just a dirty way to check it works but in the real world, it would act inconsistently depending on the fact a field is a string on the root level or n object reference due to the unobserved being wrapped by 2 different membranes.
See https://codepen.io/SimoTod/pen/qBNNQdY and click for both clone, then change foo then change test.
The first case will update the value in both the component so when you trigger the reactivity on the second one, foo updates.
The second case will update only the first component despite having the same syntax.

We can maybe open an issue first and try to get his attention before sending a PR.

**Comment by KevinBatdorf** [comment]
Okay that make sense (for the most part). I thought it was keeping the events registered but maybe not. I kept experiencing the behavior in the second example.

**Comment by KevinBatdorf** [comment]
@SimoTod I'm thinking we don't really have to wait on the clone fix before merging this in since it's essentially an Alpine bug. If people end up using this helper and running into issues we can ask them to comment on the [issue I opened](https://github.com/alpinejs/alpine/issues/835). 

The refactor PR also needs to be merged in as well, if anyone has time to take a look: https://github.com/alpine-collective/alpine-magic-helpers/pull/24

I'm ready to work on the next helper 🕶

**Comment by KevinBatdorf** [comment]
@SimoTod You probably saw it but Caleb commented on the clone issue I opened https://github.com/alpinejs/alpine/pull/859#issuecomment-720591010

>Can you paste in a bit of code that demonstrates why this is a hangup for you?
>
>I believe I NEEDED to pass the canonical element reference, so that $wire could access the ACTUAL runtime Livewire component when an Alpine component is cloned for morphdom.
>
>Has someone tested that passing just this.$el still works with Livewire's $wire across morphdom diffs?

Any initial thoughts? Otherwise I think I will have to dig into how Livewire works a bit more to make sure nothing can break. 

And maybe add a better clone test in Alpine

**Comment by SimoTod** [comment]
Yeah, I don't know if there a specific case to test but i just created an Alpine component inside a livewire component using $wire with a server side counter and it didn't break when livewire updated it.

**Comment by KevinBatdorf** [comment]
I'll try to create a livewire demo with this undo helper maybe that will give him something convenient to look at and test.

**Comment by KevinBatdorf** [comment]
This PR is good I think but the test will fail until Caleb makes the next Alpine release

**Comment by HugoDF** [comment]
@KevinBatdorf if we're using `innerText` in any of the tests, we'll need to update to use `textContent` 😄

**Comment by KevinBatdorf** [comment]
Looks like the interval test is failing. =/ I'll check it out in a bit.

**Comment by KevinBatdorf** [comment]
No `innerText` on that test (or anywhere in the repo). Maybe from other tests, Tests pass on my machine though. Any ideas?

**Comment by KevinBatdorf** [comment]
Nevermind. Updating Jest worked it out I guess. If anyone wants to check this out and approve it, it's working now with the recent release of Alpine.

Hugo, hmm I wonder if something like this (to undo state changes) would be a good addition to devtools. Maybe something users could enable under the "experimental" settings.

**Comment by SimoTod** [comment]
Just rerun it when it happens. I think it's because Javascript is not really a concurrent language so when you set a timeout, the callback could  happen 1ms after if the thread having the timeout is waiting for the priority

**Comment by KevinBatdorf** [comment]
Yeah could be. I tried running it three times. 4th time's the charm I guess :)

**Comment by KevinBatdorf** [comment]
Maybe need to update the test then though

**Comment by SimoTod** [comment]
It will lose a bit of meaning but we could allow more time instead of 100. 110 maybe would do

**Comment by KevinBatdorf** [comment]
Or maybe this?

Instead of `expect(document.querySelector('p').textContent).toEqual('1')`

```js
expect(parseInt(document.querySelector('p').textContent, 10)).toBeGreaterThan(0)
expect(parseInt(document.querySelector('p').textContent, 10)).toBeLessThan(2)
```

**Comment by SimoTod** [comment]
Do you think it would fix the issue? If yes so it's not what i though it was. The counter is just adding 1 after each seconds so checking if it's equal to 1 or between 0 and 2 should be the same I think.
The error is on the check after 3 seconds, it's expecting 4 but it's too quick and it gets 3.

**Comment by HugoDF** [comment]
> Hugo, hmm I wonder if something like this (to undo state changes) would be a good addition to devtools. Maybe something users could enable under the "experimental" settings.

🤔 might be useful but since we're sending messages back and forth, it might be easier to snapshot the state without running inside the app itself

**Comment by KevinBatdorf** [comment]
> Do you think it would fix the issue? If yes so it's not what i though it was. The counter is just adding 1 after each seconds so checking if it's equal to 1 or between 0 and 2 should be the same I think.
> The error is on the check after 3 seconds, it's expecting 4 but it's too quick and it gets 3.

Not sure actually. I thought Jest would wait a few seconds to see if it ever got to 4, but I don't understand how the timers work in Jest. I guess it's just manual. This test passes on my machine:

```js
document.body.innerHTML = `
    <div x-data="{counter: '0'}" x-init="$interval(() => counter++, 1000)">
        <p x-text="counter"></p>
    </div>
`
expect(document.querySelector('p').textContent).toEqual('0')

jest.advanceTimersByTime(5000) // I'd expect 5 iterations

await waitFor(() => {
    expect(document.querySelector('p').textContent).toEqual('1')
})
```

I think the important test is that it's actually incrementing at least once? 

Maybe just simplify to this?
```js
jest.advanceTimersByTime(2000)
expect(parseInt(document.querySelector('p').textContent, 10)).tobeGreaterThan(0)
```

**Comment by SimoTod** [comment]
It would be better to test the actual behavior. If we expect 5 and the test passes with 1, we won't catch the case where there is a bug and interval stops after the first iteration. Did you add the line to use the jest timers at the top of the file?

**Comment by KevinBatdorf** [comment]
I was just editing the existing test. Mobile now so can't easily check, but if it's there in the existing test then I didn't touch it. 

Do you think we should just increase 100 to 110? If that works then we can just do that. I don't necessarily understand what that's doing though but I can look at that another time.

**Comment by SimoTod** [comment]
If you have used the existing test, yeah there should be a line at the top to use the jest utility to mock/manipulate the timer. It's weird that it passes through, I bet that jest timer doesn't work really well. I think it will pass if we increase it, maybe even 200. As long as it less then 1000, it's fine by me

**Comment by KevinBatdorf** [comment]
Let's try 200 then and see what happens. This repo might be on hold for a bit anyway, or perhaps tied to V2 as well.

Ready for testing/approval whenever anyone has some time.

**Comment by KevinBatdorf** [comment]
Anyone have time to check on this this weekend? Otherwise I can release it separately probably.

**Review by SimoTod — APPROVED** [review]

**Pull request merged**
